### DISCUSSION TITLE
Need help testing #445

### DISCUSSION THREAD

**Discussion by L3MON4D3** [L3MON4D3/LuaSnip#519] [General]
For some details see [here](https://www.reddit.com/r/neovim/comments/wnetwo/luasnip_need_help_testing_a_biiiig_pr/?utm_source=share&utm_medium=web2x&context=3), the gist is that it's a pretty big PR, with lots of changes, and I'd love to get some real-world usage on it, if possible :D

**Comment by gegoune** [comment]
Are any actions/changes required from users apart of simply using the branch?

**Reply by L3MON4D3** [reply]
Nope, no breaking changes this time :D

**Comment by gegoune** [comment]
One thing I noticed was that packer builds `jsregexp` on every `PackerSync` with `rocks = { 'jsregexp' }` added. Is that luarocks package strictly necessary?

Other than that all snippets I tried so far work just fine.

**Reply by L3MON4D3** [reply]
Haven't experienced that yet, I added jsregexp via `use_rocks` though, maybe that's it.
jsregexp is only necessary for transformations in lsp-snippets, if you don't need that, Luasnip will work just fine without it.
Everything else working sounds good :D

**Reply by gegoune** [reply]
No idea if I need it. :) But tried using `use_rocks` and it still gets rebuilt every time I update my plugins.

**Reply by L3MON4D3** [reply]
Sounds like a `packer.nvim`-problem :/
Good to know though, for me it always worked after the initial setup, but if you have those problems, they might be common...

**Comment by leiserfg** [comment]
@L3MON4D3  what do you think of providing jsregexp via ffi as the telescope guys do with [telescope-fzf-native.nvim]( https://github.com/nvim-telescope/telescope-fzf-native.nvim/)  or even providing the pre-built libs like in [fzy-lua-native](https://github.com/romgrk/fzy-lua-native)?  I still think most of us use nvim+luajit and that will make it easier/more performant, those forced to use PUC can still use the version from rocks.

**Reply by L3MON4D3** [reply]
AFAICT they still need a build-step, that would also be the case for us, true?
If we go that way, we might as well ship `jsregexp`, compile it, and place the binary somewhere nvim can find it.
(What value would ffi provide in this configuration?)

**Reply by leiserfg** [reply]
> AFAICT they still need a build-step, that would also be the case for us, true? If we go that way, we might as well ship `jsregexp`, compile it, and place the binary somewhere nvim can find it. (What value would ffi provide in this configuration?)

If we ship a pre-built lib, using ffi makes it easier to work across distros because the lib only depends on `libc`, not on `lualib`.

**Reply by L3MON4D3** [reply]
Ahh, okay that makes sense.
But is it feasible to ship a pre-built lib? Wouldn't we need to compile for each architecture(/distro??)

**Reply by leiserfg** [reply]
Check https://github.com/romgrk/fzy-lua-native (used in  telescope-fzy-native) they do the same.

**Reply by leiserfg** [reply]
https://github.com/romgrk/fzy-lua-native/tree/master/static to be precise

**Reply by L3MON4D3** [reply]
Okay, so IIUC we have those options:
1. don't ship jsregexp, recommend install via luarock (bad, luarocks are not simple to handle via packer)
2. include jsregexp, `make` it on install (we'd have to put some more effort into the makefile, if lua(jit) is not present) (ahhhhh, here using FFI instead of the regular lua-interface would really come in handy, but since ffi is only available for luajit, I'd prefer maintaining a build-script that supports both)
3. ship (static?) libraries (can there be complications if the lua-version of the interpreter does not match the lua version used when compiling?)
4. ship (static?) libraries, call via ffi

AFAICT if we ship static libraries, the main difference between the last two would be speed, probably? There's no dependency on `libc`, nor `lualib`.
3 we could implement the quickest I think, we can switch to 4 at any time (so I'm slighly leaning toward 3 :sweat_smile:)

**Reply by leiserfg** [reply]
Theoretically building against a very old libc+liblua5.1 combo should work, we could try that and see how it goes :crossed_fingers: , Is very sad to me having to deal with distros not shipping nvim with luajit.

**Reply by L3MON4D3** [reply]
Very old for backwards compatibility?

> Is very sad to me having to deal with distros not shipping nvim with luajit

Yeah :/ It is how it is

**Reply by leiserfg** [reply]
> Very old for backwards compatibility?
> 
> > Is very sad to me having to deal with distros not shipping nvim with luajit
> 
> Yeah :/ It is how it is

Maybe it does not have to be that old, 2018 should be fine, if someone is using an older system they can build the lib themself.

**Reply by leiserfg** [reply]
Another option I was thinking of, use the lib built with `ffi` and in the case of users using `PUC` they can fall back to <https://luarocks.org/modules/colesbury/luaffi> which will remove the need for 2 versions of the lib, we just need someone with PUC to test that it actually works.

**Reply by L3MON4D3** [reply]
> Another option I was thinking of, use the lib built with ffi and in the case of users using PUC they can fall back to https://luarocks.org/modules/colesbury/luaffi which will remove the need for 2 versions of the lib, we just need someone with PUC to test that it actually works.

I can see your point, making it easy for most users, but having even more dependencies (which would again have to be installed via luarocks), I don't know...

The more I'm thinking about it, the more seems building `jsregexp` locally like the safest way to provide it.
Then we'd only have the problem with the build script having to handle both lua and luajit, but for that we could just check which lua is installed via `nvim -v`.

**Reply by L3MON4D3** [reply]
I implemented the "build-locally"-approach in `parse_from_ast`, can you see anything that might fail?
